File: parsers/cpp/src/diff.cpp
// Structural diff. Mirrors synx-core/src/diff.rs.
#include "synx/diff.hpp"

#include <algorithm>

namespace synx {

namespace {

const Value* find(const Object& obj, const std::string& key) noexcept {
    for (const auto& p : obj) {
        if (p.key == key) return &p.value;
    }
    return nullptr;
}

bool deep_equal(const Value& a, const Value& b) {
    return a.equals(b);
}

} // namespace

DiffResult diff(const Object& a, const Object& b) {
    DiffResult d;

    for (const auto& pa : a) {
        const Value* bv = find(b, pa.key);
        if (!bv) {
            d.removed.push_back(Pair{pa.key, pa.value});
        } else if (deep_equal(pa.value, *bv)) {
            d.unchanged.push_back(pa.key);
        } else {
            d.changed.push_back({pa.key, DiffChange{pa.value, *bv}});
        }
    }
    for (const auto& pb : b) {
        if (!find(a, pb.key)) {
            d.added.push_back(Pair{pb.key, pb.value});
        }
    }
    std::sort(d.unchanged.begin(), d.unchanged.end());
    return d;
}

Value diff_to_value(const DiffResult& d) {
    Object root;
    root.push_back(Pair{"added",   Value::make_object(d.added)});
    root.push_back(Pair{"removed", Value::make_object(d.removed)});

    Object changed_obj;
    for (const auto& entry : d.changed) {
        Object inner;
        inner.push_back(Pair{"from", entry.second.from});
        inner.push_back(Pair{"to",   entry.second.to});
        changed_obj.push_back(Pair{entry.first, Value::make_object(std::move(inner))});
    }
    root.push_back(Pair{"changed", Value::make_object(std::move(changed_obj))});

    Array unchanged_arr;
    unchanged_arr.reserve(d.unchanged.size());
    for (const auto& s : d.unchanged) {
        unchanged_arr.push_back(Value::make_string(s));
    }
    root.push_back(Pair{"unchanged", Value::make_array(std::move(unchanged_arr))});

    return Value::make_object(std::move(root));
}

} // namespace synx

File: parsers/dart/lib/src/diff.dart
// Structural diff. Mirrors crates/synx-core/src/diff.rs.

import 'value.dart';

class SynxDiffChange {
  final SynxValue from;
  final SynxValue to;
  const SynxDiffChange(this.from, this.to);
}

class SynxDiffEntry {
  final String key;
  final SynxDiffChange change;
  const SynxDiffEntry(this.key, this.change);
}

class SynxDiffResult {
  final SynxObject added;
  final SynxObject removed;
  final List<SynxDiffEntry> changed;
  final List<String> unchanged;
  const SynxDiffResult(this.added, this.removed, this.changed, this.unchanged);
}

SynxDiffResult diff(SynxObject a, SynxObject b) {
  final added = SynxObject();
  final removed = SynxObject();
  final changed = <SynxDiffEntry>[];
  final unchanged = <String>[];

  for (final entry in a.entries) {
    final bv = b[entry.key];
    if (bv == null && !b.contains(entry.key)) {
      removed.set(entry.key, entry.value);
    } else if (entry.value == bv) {
      unchanged.add(entry.key);
    } else {
      changed.add(SynxDiffEntry(entry.key, SynxDiffChange(entry.value, bv!)));
    }
  }
  for (final entry in b.entries) {
    if (a[entry.key] == null && !a.contains(entry.key)) {
      added.set(entry.key, entry.value);
    }
  }
  unchanged.sort();
  return SynxDiffResult(added, removed, changed, unchanged);
}

SynxValue diffToValue(SynxDiffResult d) {
  final root = SynxObject();
  root.set('added', synxObject(d.added));
  root.set('removed', synxObject(d.removed));

  final changed = SynxObject();
  for (final e in d.changed) {
    final inner = SynxObject();
    inner.set('from', e.change.from);
    inner.set('to', e.change.to);
    changed.set(e.key, synxObject(inner));
  }
  root.set('changed', synxObject(changed));

  final arr = d.unchanged.map((s) => synxString(s)).toList();
  root.set('unchanged', synxArray(arr));
  return synxObject(root);
}
